src/controllers/hid/hidiothread.cpp

```diff
@@ -0,0 +1,236 @@
+    // This eexecute callback function in JavaScript mapping and print to stdout in case of --controllerDebug
+    emit receive(QByteArray(reinterpret_cast<const char*>(pCurrentBuffer),
+                         bytesRead),
+            mixxx::Time::elapsed());
+}
+
+QByteArray HidIoThread::getInputReport(unsigned int reportID) {
+    auto startOfHidGetInputReport = mixxx::Time::elapsed();
+    auto lock = lockMutex(&m_HidDeviceMutex);
+    int bytesRead;
+
+    m_pPollData[m_pollingBufferIndex][0] = reportID;
+    // FIXME: implement upstream for hidraw backend on Linux
+    // https://github.com/libusb/hidapi/issues/259
+    bytesRead = hid_get_input_report(m_pHidDevice, m_pPollData[m_pollingBufferIndex], kBufferSize);
+    qCDebug(m_logInput) << bytesRead << "bytes received by hid_get_input_report"
+                        << m_deviceInfo.formatName() << "serial #"
+                        << m_deviceInfo.serialNumber()
+                        << "(including one byte for the report ID:"
+                        << QString::number(static_cast<quint8>(reportID), 16)
+                                   .toUpper()
+                                   .rightJustified(2, QChar('0'))
+                        << ") - Needed: "
+                        << (mixxx::Time::elapsed() - startOfHidGetInputReport)
+                                   .formatMicrosWithUnit();
+
+    if (bytesRead <= kReportIdSize) {
+        // -1 is the only error value according to hidapi documentation.
+        // Otherwise minimum possible value is 1, because 1 byte is for the reportID,
+        // the smallest report with data is therefore 2 bytes.
+        DEBUG_ASSERT(bytesRead <= kReportIdSize);
+        return QByteArray();
+    }
+
+    // Convert array of bytes read in a JavaScript compatible return type, this is returned as deep-copy, for thread safety.
+    // For compatibility with HidController::processInputReport, the reportID prefix is included added here
+    return QByteArray(reinterpret_cast<char*>(m_pPollData[m_pollingBufferIndex]), bytesRead);
+}
+
+void HidIoThread::sendOutputReport(const QByteArray& data, unsigned int reportID) {
+    {
+        auto lock = lockMutex(&m_HidDeviceMutex);
+        if (m_outputReports.find(reportID) == m_outputReports.end()) {
+            std::unique_ptr<HidIoReport> pNewOutputReport;
+            m_outputReports[reportID] = std::make_unique<HidIoReport>(
+                    reportID, m_pHidDevice, std::move(m_deviceInfo));
+        }
+        // SendOutputReports executes a hardware operation, which take several milliseconds
+        m_outputReports[reportID]->sendOutputReport(data);
+    }
+    // Ensure that all InputReports are read from the ring buffer, before the next OutputReport blocks the IO again
+    poll(); // Polling available Input-Reports is a cheap software only operation, which takes insignificiant time
```

Comment:
Do we want to poll also before the thread is blocked?

Reply:
It's only important, that the readout is done before the event queue get control. Because the next event can be a time consuming OutputReport too.